pandas/tests/reshape/merge/test_merge.py

```diff
@@ -1442,3 +1442,36 @@ def test_merge_series(on, left_on, right_on, left_index, right_index, nm):
         with pytest.raises(ValueError, match='a Series without a name'):
             result = pd.merge(a, b, on=on, left_on=left_on, right_on=right_on,
                               left_index=left_index, right_index=right_index)
+
+
+@pytest.mark.parametrize("col1, col2, suffixes, expected_cols", [
+    (0, 0, ("", "_dup"), ["0", "0_dup"]),
+    (0, 0, (None, "_dup"), ["0", "0_dup"]),
```

Comment:
> Even that type isn't quite right... Since (False, False) is apparently valid.

Yeah, it looks like pretty much anything is actually valid right now, e.g.
```python
In [8]: df.merge(df, left_index=True, right_index=True, suffixes=(np.sum, pd.DataFrame))
Out[8]:
   0<function sum at 0x000002550602C840>  0<class 'pandas.core.frame.DataFrame'>
0                                      1                                       1

In [9]: df.merge(df, left_index=True, right_index=True, suffixes=(list('abc'), pd.Series(list('xyz'))))
Out[9]:
   0['a', 'b', 'c']  00    x\n1    y\n2    z\ndtype: object
0                 1                                       1                                   1                                       1
```
Looks to be caused by `items_overlap_with_suffix` just hitting whatever is passed with a `format`:
https://github.com/pandas-dev/pandas/blob/08f92c4c86fbb4e03d030f24ea66402f6deee6df/pandas/core/internals/managers.py#L1959-L1985

> IIUC, @simonjayhawkins expected for (None, '_dup') to not cast the original input to a string.

Yeah, I overlooked this earlier but agree; it makes sense to me that `None` should not cast to string and instead retain the original type of the column label in question.

Reply:
thanks! @TomAugspurger @jschendel i overlooked it... i changed the code a bit, and added several more test cases
And @jschendel I also feel like we should add some restrictions to the suffix type, apparently, it looks wired if any type is given as suffixes... is it worth it if i open another PR to address it?